cygwin: add as a cross-compilation target, and get bash.exe to compile

puffnfresh opened the pull request:
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RossComputerGuy commented on lib/systems/examples.nix:
```diff
@@ -325,6 +325,11 @@ rec {
     libc = "ucrt"; # This distinguishes the mingw (non posix) toolchain
   };
 
+  cygwin = {
```
Probably should call this `x86_64-cygwin`.

RossComputerGuy commented on pkgs/os-specific/windows/cygwin/headers.nix:
```diff
@@ -0,0 +1,40 @@
+, mingw_w64_headers
+}:
+
+stdenvNoCC.mkDerivation rec {
```
Please use `finalAttrs:`

RossComputerGuy commented on pkgs/os-specific/windows/cygwin/headers.nix:
```diff
@@ -0,0 +1,40 @@
+    cp -r winsup/cygwin/include/* $out/include/
+  '';
+
+  meta = {
+    platforms = lib.platforms.windows;
+  };
```
Could be folded to just `meta.platforms` but it might be good to fill in the rest of the typical metadata.

alyssais replied:
This had been marked as resolved but doesn't seem to have actually been?

RossComputerGuy commented on pkgs/by-name/co/cocom-tool-set/package.nix:
```diff
@@ -0,0 +1,30 @@
+  fetchgit,
+  bison,
+}:
+stdenv.mkDerivation rec {
```
Please use `finalAttrs:` instead of `rec`

RossComputerGuy commented on pkgs/development/libraries/readline/8.2.nix:
```diff
@@ -3,7 +3,7 @@
 , ncurses, termcap
 , curses-library ?
-    if stdenv.hostPlatform.isWindows
+    if stdenv.hostPlatform.isMinGW
```
What's the exact reason for changing from checking for windows in general to mingw? Would `isWindows || isMinGW` work better?

puffnfresh replied:
`isWindows` includes Cygwin. Under Cygwin we want to use ncurses, not termcap.

RossComputerGuy commented on pkgs/development/libraries/readline/8.2.nix:
```diff
@@ -35,13 +35,15 @@ stdenv.mkDerivation rec {
   ]
   ++ upstreamPatches
-  ++ lib.optionals stdenv.hostPlatform.isWindows [
+  ++ lib.optionals stdenv.hostPlatform.isMinGW [
```
What's the exact reason for changing from checking for windows in general to mingw? Would `isWindows || isMinGW` work better?

puffnfresh replied:
We don't want to apply these patches under Cygwin. Cywin provides a POSIX API under Windows, so we want to use the normal POSIX calls which readline would usually make.

emilazy replied:
I guess this could be something like `(stdenv.hostPlatform.isWindows && !stdenv.hostPlatform.isUnix)` if we wanted to mark Cygwin as the technically‐correct Windows + Unix and confuse everyone for the rest of time?

puffnfresh replied:
@emilazy yeah really good point, MSVC is on its way. I actually like the suggestion of `isWindows && !isUnix`.

emilazy replied:
Just be prepared for the fact that there’s probably a lot of things in the tree that assume the two are disjoint. But it’d be great to deconfuse that along the way.

RossComputerGuy reviewed (changes requested):
Please squash specific changes amongst your commits. You can make several commits and use `git rebase -i` to do that. Also, please format your nix files so CI can pass.

Artturin commented:
> Please squash specific changes amongst your commits. You can make several commits and use `git rebase -i` to do that. Also, please format your nix files so CI can pass.

Check out git-absorb for this @puffnfresh 
You can `git reset --soft HEAD and then run it to absorb the changes

alyssais commented on pkgs/development/libraries/ncurses/default.nix:
```diff
@@ -99,6 +99,8 @@ stdenv.mkDerivation (finalAttrs: {
     CFLAGS=-D_XOPEN_SOURCE_EXTENDED
   '';
 
+  hardeningDisable = lib.optionals stdenv.hostPlatform.isCygwin [ "fortify" ];
```
If this needs to be set for every package, why not globally disable fortify for cygwin?

puffnfresh replied:
@alyssais I wanted to, but I'm not sure where the best place to do that would be. Should it be an additional condition in the default value of `defaultHardeningFlags` in `pkgs/build-support/bintools-wrapper`?

Ericson2314 replied:
This will also have the added benefit of making it not a mass rebuild, I think

puffnfresh replied:
I've moved this to `bintools-wrapper` and added a few things to avoid the mass rebuild.

Ericson2314 replied:
Great!

alyssais replied:
The commits that set it individually for packages need to be updated then.

alyssais reviewed (changes requested):
This needs to be more than two commits.  One logical change at a time, that gets it a bit further.

puffnfresh committed:
cygwin: init as a target toolchain

puffnfresh committed:
cygwin: add pre-libc headers

puffnfresh committed:
cygwin: remove old w32api, it's in mingw-w64

puffnfresh committed:
cocom-tool-set: init at 0.996

puffnfresh committed:
cygwin: package newlib-cygwin as the libc

puffnfresh committed:
cygwin: create libc from joined newlib-cygwin and mingw_w64

puffnfresh committed:
readline: fix compilation under Cygwin

puffnfresh committed:
ncurses: disable fortify under Cygwin

puffnfresh committed:
bash: fix compilation under Cygwin

puffnfresh committed:
cygwin: include libcygwin1.dll in DLL search path

puffnfresh commented:
> This needs to be more than two commits. One logical change at a time, that gets it a bit further.

I've split this into 10 commits now. Also resolved conflicts.

Ericson2314 commented on lib/systems/parse.nix:
```diff
@@ -444,7 +444,7 @@ rec {
     "2" = # We only do 2-part hacks for things Nix already supports
       if elemAt l 1 == "cygwin"
-        then { cpu = elemAt l 0;                      kernel = "windows";  abi = "cygnus";   }
+        then mkSkeletonFromList [ (elemAt l 0) "pc" "cygwin" ]
```
This change is purely a matter of style, not behavior, right?

(I kinda wonder if we can just drop this case, as the older cygwin stuff is bit-rotted for years at this point.)

puffnfresh replied:
I'm in favour of getting rid of all the older Cygwin support, since it just doesn't work anymore (I tried).

Ericson2314 replied:
Yeah it's well over a decade old

puffnfresh replied:
I had a go at removing the hack altogether: https://github.com/puffnfresh/nixpkgs/commit/16e9e8bbb798604a10e8e29706b0a94146b04ba1

It feels a little bit wrong to add Cygwin as a "kernel" but it does seem to work. Let me know what you think.

Ericson2314 commented on pkgs/os-specific/windows/cygwin/after-autogen.patch:
Could we instead patch `configure.ac` and regen with `autoreconfHook`?

puffnfresh replied:
Maybe. This was copied from Microsoft's [Windows-on-ARM-Experiments repository](https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/blob/c50f0d8e7fdc531344a282ef58bd4eb653658b3f/patches/cygwin/0002-after-autogen.patch). Not sure why it was done that way.

Ericson2314 replied:
oh probably because they are lazy af lol

alyssais replied:
Why copied and not fetched?  If you used fetchpatch you could even use `extraPrefix` to make it so you can just put it in `patches`.

puffnfresh committed:
bintools-wrapper: don't set fortify on Cygwin

The size of various things don't line up when enabling fortification.
This might be able to be fixed by looking at the Cywin builds, because
I think they don't generally disable fortify.

puffnfresh committed:
bash: set empty makeFlags to avoid mass rebuild

puffnfresh committed:
mingw-w64: allow Cygwin as a crt

Ericson2314 commented:
Oh is this `ofborg-eval-darwin` failure spurious?

Ericson2314 commented:
OK this is evaluating now!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/dropping-cygwin-support/3425/19

alyssais commented on pkgs/development/compilers/gcc/common/configure-flags.nix:
```diff
@@ -242,6 +242,10 @@ let
       "--disable-werror"
     ]
+    ++ lib.optionals targetPlatform.isCygwin [
+      # doesn't cross-compile under Cygwin
```
Do you mean "for Cygwin" (when running under some other platform)? Otherwise the check is wrong.

alyssais commented on pkgs/os-specific/windows/cygwin/headers.nix:
```diff
@@ -0,0 +1,40 @@
+  };
+
+  patches = [
+    (fetchpatch {
```
This is a static file, so there's no need for patch normalization and fetchurl can be used.

alyssais commented on pkgs/by-name/co/cocom-tool-set/package.nix:
```diff
@@ -0,0 +1,28 @@
+
+  nativeBuildInputs = [ bison ];
+
+  patches = [ ./remove-lto.patch ];
```
This could do with some explanation, ideally in the patch file.

alyssais commented on pkgs/by-name/co/cocom-tool-set/package.nix:
```diff
@@ -0,0 +1,28 @@
+    url = "https://git.code.sf.net/p/cocom/git";
+    sha256 = "sha256-utLafkznMC4LrZgF6vKehtIGMwNMwLP9M9Nwu/RyWio=";
+    rev = "64ee80224aa13f9944d439f3f90862ca76158705";
+  };
```
Why not the [release tarball](https://sourceforge.net/projects/cocom/files/cocom%20sources/cocom-0.996/)?

alyssais commented on pkgs/os-specific/windows/cygwin/default.nix:
```diff
@@ -0,0 +1,65 @@
+    "--enable-static"
+    "--disable-dumper"
+    "--with-cross-bootstrap"
+    "ac_cv_prog_CC=gcc"
```
Why?  This looks like it's hardcoding an assumption that the compiler is GCC.

alyssais commented on pkgs/os-specific/windows/mingw-w64/default.nix:
```diff
@@ -33,6 +33,22 @@ stdenv.mkDerivation {
+    ln -fs w32api/libpsapi.a .
+    ln -fs w32api/libuserenv.a .
+    ln -fs w32api/libnetapi32.a .
+    ln -fs w32api/libdbghelp.a .
```
Would w32api/*.a be wrong?

Changing directory without changing back at the end also makes me worry about fixupPhase not working.

alyssais commented on pkgs/os-specific/windows/cygwin/default.nix:
```diff
@@ -10,56 +10,77 @@
+    ];
+  };
+in
+# TODO: Is there something like nix-support which would achieve this better?
```
Could you just link the appropriate paths in postInstall in the main derivation?

alyssais commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -474,6 +474,7 @@ stdenvNoCC.mkDerivation {
         echo addToSearchPath "LINK_DLL_FOLDERS" "${cc_solib}/lib" > $out
         echo addToSearchPath "LINK_DLL_FOLDERS" "${cc_solib}/lib64" >> $out
         echo addToSearchPath "LINK_DLL_FOLDERS" "${cc_solib}/lib32" >> $out
+        echo addToSearchPath "LINK_DLL_FOLDERS" "${libc_bin}/bin" > $out
```
It's a bit surprising that it's in bin — is that definitely right?

alyssais commented on pkgs/shells/bash/5.nix:
```diff
@@ -107,6 +107,8 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
 
+  makeFlags = [ ];
```
Please don't make the code more complicated just to avoid a mass rebuild.  Mass rebuilds are fine.  The bash changes can do to staging separately from the rest of this if you want.

corngood replied:
> Please don't make the code more complicated just to avoid a mass rebuild.

I agree with this sentiment, but I'm struggling with how to handle in my update of this PR.

I don't want to put the whole thing through staging, because I want to prove that it has no impact on other platforms, but sometimes these attribute changes are coupled with more substantial changes.

My instinct is to try to avoid rebuilds in master, and then do a cleanup change in staging afterward.

Anyone know if there's been any thought put into this as a systemic problem?

philiptaron replied:
The way to do this is to base on `$(git merge-base master staging)`, so the commit upon which this branch is based is in `master` but the PR targets `staging`. That way, people can check out the branch if they want to not rebuild everything.

I'm going to echo Alyssa's advice: just do the right change, and let the rebuilds fall out where they may.

corngood replied:
I just want to be sure that I'm not introducing unexpected changes to other platforms.  If I'm careful to avoid rebuilds, I can just run a nixpkgs-review against master and see 0 package changes on e.g. linux.

As long as I have changes to something like bash in there, it gets very difficult to identify unintended changes.  Or at least I haven't found a good way to do it.

If I have to, I can always keep my minimal rebuild branch locally, and test that before squashing the cleanups and making a PR to staging.
